captum/optim/_utils/reducer.py

```diff
@@ -0,0 +1,192 @@
+        orig_shape = x.shape
+        return func(x.reshape([-1, x.shape[-1]])).reshape(list(orig_shape[:-1]) + [-1])
+
+    def fit_transform(
+        self, x: Union[torch.Tensor, np.ndarray], reshape: bool = True
+    ) -> Union[torch.Tensor, np.ndarray]:
+        """
+        Perform dimensionality reduction on an input tensor or NumPy array.
+        """
+
+        is_tensor = torch.is_tensor(x)
+
+        if is_tensor:
+            if x.dim() == 3 and reshape:
+                x = x.permute(2, 1, 0)
+            elif x.dim() > 3 and reshape:
+                permute_vals = [0] + list(range(x.dim()))[2:] + [1]
+                x = x.permute(*permute_vals)
+        else:
+            if x.ndim == 3 and reshape:
+                x = x.transpose(2, 1, 0)
+            elif x.ndim > 3 and reshape:
+                permute_vals = [0] + list(range(x.ndim))[2:] + [1]
+                x = x.transpose(*permute_vals)
+
+        x_out = ChannelReducer._apply_flat(self._reducer.fit_transform, x)
+
+        if is_tensor:
+            x_out = torch.as_tensor(x_out, device=x.device)
+            if x.dim() == 3 and reshape:
+                x_out = x_out.permute(2, 1, 0)
+            elif x.dim() > 3 and reshape:
+                permute_vals = (
+                    [0]
+                    + [x.dim() - 1]
+                    + list(range(x.dim()))[1 : len(list(range(x.dim()))) - 1]
+                )
+                x_out = x_out.permute(*permute_vals)
+        else:
+            if x.ndim == 3 and reshape:
+                x_out = x_out.permute(2, 1, 0)
+            elif x.ndim > 3 and reshape:
+                permute_vals = (
+                    [0]
+                    + [x.ndim - 1]
+                    + list(range(x.ndim))[1 : len(list(range(x.ndim))) - 1]
+                )
+                x_out = x_out.transpose(*permute_vals)
+
+        return x_out
+
+    def __getattr__(self, name: str) -> Any:
+        if name in self.__dict__:
+            out = self.__dict__[name]
+        elif name + "_" in self._reducer.__dict__:
+            out = self._reducer.__dict__[name + "_"]
+        if type(out) == np.ndarray:
+            out = torch.as_tensor(out)
+        return out
+
+    def __dir__(self) -> List:
+        dynamic_attrs = [
+            name[:-1]
+            for name in dir(self._reducer)
+            if name[-1] == "_" and name[0] != "_"
+        ]
+
+        return (
+            list(ChannelReducer.__dict__.keys())
+            + list(self.__dict__.keys())
+            + dynamic_attrs
+        )
+
+
+def posneg(x: torch.Tensor, dim: int = 0) -> torch.Tensor:
+    """
+    Hack that makes a matrix positive by concatination in order to simulate
+    one-sided NMF with regular NMF
+    """
+
+    return torch.cat(
+        [torch.max(x, torch.full_like(x, 0)), torch.max(-x, torch.full_like(x, 0))],
+        dim=dim,
+    )
+
+
+def nchannels_to_rgb(x: torch.Tensor, warp: bool = True) -> torch.Tensor:
+    """

```

Comment:
@NarineK Yeah, but it'll be a lot easier to use as a function in some cases. So, maybe I can create a wrapper class for `transform.py`? Should I leave the function in `utils/common.py` or should I also move it to `transform.py`?

Reply:
I see  - If we have it as a transform how cumbersome will it be to use ? If we need a wrapper, we can keep the wrapper in the `_utils/image/common.py` and the transformation piece in the transform.py (if it is possible to represent as a transformation) ?

Reply:
The Lucid equivalent of nchannels_to_rgb is built into their `show()` function but sometimes it's called separately, and it's often used for displaying weights. If we make it class, users will have to use two lines of code instead of one to replicate the behavior. I also think don't Lucid ever uses it as a transform for optimizing model inputs with.

Reply:
Sounds good, we can have the wrapper one line function to perform that transformation in the common.py.

Reply:
I know PyTorch has some of it's layer classes as wrappers for the functional equivalents, so I think it makes more sense to have the Class call the function as opposed to the other way around.